### PR TITLE
Support array attribute in stablehlo composite

### PR DESCRIPTION
**Pull request by chunnienc**
Support translating `list[int]`, `list[float]`, `list[bool]` python value to array attribute in composite.

Internal reference: https://b.corp.google.com/issues/331634303

### CODE REVIEW

File: torch_xla/csrc/runtime/stablehlo_composite_helper.cc [hunk 1]
```diff
@@ -190,35 +190,64 @@ class BuildStableHLOCompositePass : public mlir::OperationPass<mlir::ModuleOp> {
...
+            el_type = json::value_t::number_integer;
+          }
+          if (common_el_type.has_value() && common_el_type.value() != el_type) {
+            return mlir::failure();
+          }
+          common_el_type = el_type;
+        }
+        if (!common_el_type.has_value()) {
+          return builder.getArrayAttr({});
+        }
```
Comment (GleasonK) [inline torch_xla/csrc/runtime/stablehlo_composite_helper.cc:220]: Control flow here was difficult to understand. How about:

```
if (json_value.empty()) return builder.getArrayAttr({});
auto headType = json_value[0].type();
bool is_homogeneous = llvm::all_of(json_value,
  [&](auto & el){ return el.type() == headType; });
if (!is_homogeneous) emitError(builder.getUnknownLocation(), ""invalid JSON to MLIR, arrays must be homogeneous");
switch (headType.value()) {
  // ....
```

File: torch_xla/csrc/runtime/stablehlo_composite_helper.cc [hunk 2]
```diff
@@ -190,35 +190,64 @@ class BuildStableHLOCompositePass : public mlir::OperationPass<mlir::ModuleOp> {
...
+            return builder.getDenseI64ArrayAttr(
+                json_value.template get<llvm::SmallVector<int64_t>>());
+          case json::value_t::number_float:
+            return builder.getDenseF32ArrayAttr(
+                json_value.template get<llvm::SmallVector<float>>());
+          case json::value_t::boolean:
+            return builder.getDenseBoolArrayAttr(
+                json_value.template get<llvm::SmallVector<bool>>());
+          default:
+            return mlir::failure();
```
Comment (GleasonK) [inline torch_xla/csrc/runtime/stablehlo_composite_helper.cc:232]: emitError(builder.getUnknownLocation(), ""invalid JSON to MLIR, invalid array type arrays must be 1-D homogeneous arrays of supported primitive types");

File: torch_xla/csrc/runtime/stablehlo_composite_helper.cc [hunk 3]
```diff
@@ -190,35 +190,64 @@ class BuildStableHLOCompositePass : public mlir::OperationPass<mlir::ModuleOp> {
...
+        }
+      }
+      default:
+        return mlir::failure();
```
Comment (GleasonK) [inline torch_xla/csrc/runtime/stablehlo_composite_helper.cc:236]: emitError(builder.getUnknownLocation(), "invalid JSON to MLIR, arrays must be homogeneous");

Comment (GleasonK) [reply torch_xla/csrc/runtime/stablehlo_composite_helper.cc:236]: consider an error helper function which prepends the message to let you know where the error is coming from, or thread through the custom_call we're fixing up and use `custom_call->emitError`.

File: torch_xla/csrc/runtime/stablehlo_composite_helper.cc (hunk 1 above)
Comment (GleasonK) [reply torch_xla/csrc/runtime/stablehlo_composite_helper.cc:220]: Also I'm on board with heterogeneous support, but I think it's not needed yet. Can be accomplished via a recursive call to this function.

**Review by GleasonK — APPROVED** [review]
LGTM. A few minor suggestions including error messages

**Review by lsy323 — APPROVED** [review]

File: torch_xla/csrc/runtime/stablehlo_composite_helper.cc (hunk 3 above)
Comment (chunnienc) [reply torch_xla/csrc/runtime/stablehlo_composite_helper.cc]: For now I pass the custom_call op as a arg to the build attr method, for emitting the error.

File: torch_xla/csrc/runtime/stablehlo_composite_helper.cc [hunk 4]
```diff
@@ -190,35 +190,74 @@ class BuildStableHLOCompositePass : public mlir::OperationPass<mlir::ModuleOp> {
...
+          return get_json_type(el) == head_type;
+        });
+        if (!is_homogeneous) {
+          return op->emitError()
+                 << "invalid JSON to MLIR, arrays must be homogeneous";
+        }
+
+        switch (head_type) {
+          case json::value_t::number_integer:
+            return builder.getDenseI64ArrayAttr(
```
Comment (GleasonK) [inline torch_xla/csrc/runtime/stablehlo_composite_helper.cc:229]: I'm just remembering that we don't have general serialization support for DenseArray types. Would recommend using 1-D DenseIntOrFpElements for now:

```
// I64
builder.getI64TensorAttr(values);

//float
return DenseFPElementsAttr::get(RankedTensorType::get(static_cast<float>(values.size()), builder.getF32Type()), values);

//bool
return DenseIntElementsAttr::get(RankedTensorType::get(static_cast<int64_t>(values.size()), builder.getI1Type()), values);
```

where `values` is the small vector cast of JSON you have here.

Comment (chunnienc) [reply torch_xla/csrc/runtime/stablehlo_composite_helper.cc:229]: Confirm: does this look good to you?
```
module @IrToHlo.11 attributes {mhlo.cross_program_prefetches = [], mhlo.is_dynamic = false, mhlo.use_auto_spmd_partitioning = false} {
  func.func @main(%arg0: tensor<5x5xf32>, %arg1: tensor<5x5xf32>) -> tensor<5x5xf32> {
    %0 = stablehlo.composite "test.add" %arg1, %arg0 {composite_attributes = {bool_arr = dense<[true, false]> : tensor<2xi1>, float_arr = dense<[1.000000e+00, 1.100000e+00, 1.200000e+00]> : tensor<3xf32>, int_arr = dense<[1, 2, 3]> : tensor<3xi64>}, decomposition = @test.add.impl} : (tensor<5x5xf32>, tensor<5x5xf32>) -> tensor<5x5xf32>
    return %0 : tensor<5x5xf32>
  }
  func.func private @test.add.impl(%arg0: tensor<5x5xf32>, %arg1: tensor<5x5xf32>) -> tensor<5x5xf32> {
    %0 = stablehlo.add %arg0, %arg1 : tensor<5x5xf32>
    return %0 : tensor<5x5xf32>
  }
}
```

**Review by GleasonK — APPROVED** [review]

**Pull request merged**